File: src/com/rian/difficultycalculator/beatmap/timings/TimingControlPointManager.java
package com.rian.difficultycalculator.beatmap.timings;

/**
 * A manager for timing control points.
 */
public class TimingControlPointManager extends ControlPointManager<TimingControlPoint> {
    public TimingControlPointManager() {
        super(new TimingControlPoint(0, 1000, 4));
    }

    /**
     * Copy constructor.
     *
     * @param source The source to copy from.
     */
    private TimingControlPointManager(TimingControlPointManager source) {
        super(source.defaultControlPoint);

        for (TimingControlPoint point : source.controlPoints) {
            controlPoints.add(point.deepClone());
        }
    }

    @Override
    public TimingControlPoint controlPointAt(double time) {
        return binarySearchWithFallback(time, controlPoints.size() > 0 ? controlPoints.get(0) : defaultControlPoint);
    }

    @Override
    public TimingControlPointManager deepClone() {
        return new TimingControlPointManager(this);
    }
}

File: src/com/rian/osu/beatmap/timings/TimingControlPointManager.kt
package com.rian.osu.beatmap.timings

/**
 * A manager for [TimingControlPoint]s.
 */
class TimingControlPointManager : ControlPointManager<TimingControlPoint>(
    DEFAULT_TIMING_CONTROL_POINT
) {
    override fun controlPointAt(time: Double) =
        binarySearchWithFallback(time, controlPoints.getOrNull(0) ?: defaultControlPoint)

    companion object {
        val DEFAULT_TIMING_CONTROL_POINT = TimingControlPoint(0.0, 1000.0, 4)
    }
}
